
# AOS_PS1_phosim_run

Update file paths in cmd files  to those relative to the `AOS/DM-28360/` repo, which 
is considered the `root_dir`

In [31]:
def readFile(pathToFile, headerString = '#'):
    ''' Read file line-by-line  as two lists,
    one containing the header lines, 
    the other the content lines 
    
    Parameters:
    -----------
    pathToFile: `str`
        absolute path to file 
        
    Returns:
    ----------
    `list`, `list` 
        lists containing the header lines 
        and the content lines 
    '''
    with open(pathToFile) as f:
        allLines = f.readlines()

    headerLines=[]
    contentLines=[]

    # store the header part and content separately 
    for line in allLines:
        if line.startswith('#'):
            headerLines.append(line)
        else:
            contentLines.append(line)

    return headerLines, contentLines 

def update_cmd_file_paths(bkgnd ='qck', pert='05',
                          cmd_dir = '/project/scichris/aos/AOS/DM-28360',
                          mirror_dir = '/project/scichris/aos/AOS/DM-28360',
                          cmd_file=None
                         ):
    if cmd_file == None:    
        cmd_file = f'{bkgnd}BkgndPert{pert}.cmd'
        
    path_to_cmd = os.path.join(cmd_dir, cmd_file)
    surfacemap_dir = os.path.join(f'imgCloseLoop_0-{pert}','pert','iter0')
    
    header, content = readFile(path_to_cmd)  
    #print(content)
    new_content = content.copy()
    for i in range(len(content)):
        line = content[i]
        newline = line
        print(line)
        if line.startswith('surfacemap'):

            print(line) 
            # split the original line 
            splitline = line.split()
            mirror_file = splitline[2].split('/')[-1]
            path_to_mirror = os.path.join(mirror_dir, surfacemap_dir, mirror_file)
            print(path_to_mirror)
            # replace that with a new line with an 
            # updated path 
            newSplitLine = splitline[:2] # "eg. surfacemap 0"
            # eg. /project/scichris/aos/AOS/DM-28360/imgCloseLoop_0-00/iter0/pert/M1res.txt
            print(newSplitLine)
            newSplitLine.append(path_to_mirror)
            newSplitLine.append(splitline[-1]) # eg. 1
            newline = ' '.join(newSplitLine)
            newline += ' \n'
            print('-->', newline)
        new_content[i] = newline
        return new_content
        
def write_to_file(out_file, content):
    with open(out_file, "w") as output:
        for line in content:
            output.write(line)

In [37]:

def update_cmd_mirror_paths(bkgnd, pert, cmd_dir, mirror_dir):
    

    #cmd_file = None
    #bkgnd='qck'
    #pert='00'
    #if cmd_file == None:    
    cmd_file = f'{bkgnd}BkgndPert{pert}.cmd'

    path_to_cmd = os.path.join(cmd_dir, cmd_file)
    surfacemap_dir = os.path.join(f'imgCloseLoop_0-{pert}','pert','iter0')

    header, content = readFile(path_to_cmd)  
    #print(content)

    new_content = content.copy()
    for i in range(len(content)):
        line = content[i]
        newline = line
        if line.startswith('surfacemap'):
            print(line) 
            # split the original line 
            splitline = line.split()
            mirror_file = splitline[2].split('/')[-1]
            path_to_mirror = os.path.join(mirror_dir, surfacemap_dir, mirror_file)
            #print(path_to_mirror)
            # replace that with a new line with an 
            # updated path 
            newSplitLine = splitline[:2] # "eg. surfacemap 0"
            # eg. /project/scichris/aos/AOS/DM-28360/imgCloseLoop_0-00/iter0/pert/M1res.txt
            #print(newSplitLine)
            newSplitLine.append(path_to_mirror)
            newSplitLine.append(splitline[-1]) # eg. 1
            newline = ' '.join(newSplitLine)
            newline += ' \n'
            print('-->', newline)
        new_content[i] = newline
    
    return new_content



In [41]:
suffix = 'hyak'
bkgnd='no'
for pert in ['00','05']:
    new_content = update_cmd_mirror_paths(bkgnd, pert, cmd_dir, mirror_dir)
    cmd_file = f'{bkgnd}BkgndPert{pert}.cmd'
    new_cmd_file = f'{bkgnd}BkgndPert{pert}_{suffix}.cmd'
    out_file = os.path.join(cmd_dir, new_cmd_file)
    print(f'Updated {cmd_file} surfacemap paths, saved {out_file}')
    write_to_file(out_file, new_content)

surfacemap 0 /project/scichris/aos/perturbations/imgCloseLoop_0-00/iter0/pert/M1res.txt 1 

--> surfacemap 0 /gscratch/astro/suberlak/aos/AOS/DM-28360/imgCloseLoop_0-00/pert/iter0/M1res.txt 1 

surfacemap 2 /project/scichris/aos/perturbations/imgCloseLoop_0-00/iter0/pert/M3res.txt 1 

--> surfacemap 2 /gscratch/astro/suberlak/aos/AOS/DM-28360/imgCloseLoop_0-00/pert/iter0/M3res.txt 1 

surfacemap 1 /project/scichris/aos/perturbations/imgCloseLoop_0-00/iter0/pert/M2res.txt 1 

--> surfacemap 1 /gscratch/astro/suberlak/aos/AOS/DM-28360/imgCloseLoop_0-00/pert/iter0/M2res.txt 1 

Updated noBkgndPert00.cmd surfacemap paths, saved /project/scichris/aos/AOS/DM-28360/noBkgndPert00_hyak.cmd
surfacemap 0 /project/scichris/aos/imgCloseLoop/iter0/pert/M1res.txt 1 

--> surfacemap 0 /gscratch/astro/suberlak/aos/AOS/DM-28360/imgCloseLoop_0-05/pert/iter0/M1res.txt 1 

surfacemap 2 /project/scichris/aos/imgCloseLoop/iter0/pert/M3res.txt 1 

--> surfacemap 2 /gscratch/astro/suberlak/aos/AOS/DM-28360/img

**Now** it does work correctly for `qckBkgndPert05.cmd` ! 

In [39]:
import os
 
cmd_dir ='/project/scichris/aos/AOS/DM-28360' # path where cmd file is found
mirror_dir = '/gscratch/astro/suberlak/aos/AOS/DM-28360' # path where the mirror files are 
#new_dir= "/gscratch/astro/suberlak/aos/AOS/DM-28360" # "/project/scichris/aos/AOS/DM-28360"
suffix = 'hyak' # 'NCSA'
bkgnd='qck'
for pert in ['00','05']:
    
    
    new_content = update_cmd_file_paths(bkgnd=bkgnd, pert=pert, 
                                        cmd_dir=cmd_dir, 
                                        mirror_dir=mirror_dir)

    new_cmd_file = f'{bkgnd}BkgndPert{pert}_{suffix}.cmd'
    out_file = os.path.join(cmd_dir, new_cmd_file)

    print(f'Updated {cmd_file} surfacemap paths to {new_cmd_file}, as {out_file}')
    write_to_file(out_file, new_content)

quickbackground

Updated qckBkgndPert00.cmd surfacemap paths to qckBkgndPert00_hyak.cmd, as /project/scichris/aos/AOS/DM-28360/qckBkgndPert00_hyak.cmd
quickbackground

Updated qckBkgndPert00.cmd surfacemap paths to qckBkgndPert05_hyak.cmd, as /project/scichris/aos/AOS/DM-28360/qckBkgndPert05_hyak.cmd


In [28]:
new_content

['quickbackground\n',
 'raydensity 0\n',
 'perturbationmode 1\n',
 'trackingmode 0\n',
 'cleartracking\n',
 'clearclouds\n',
 'lascatprob 0.0\n',
 'contaminationmode 0\n',
 'diffractionmode 1\n',
 'straylight 0\n',
 'detectormode 0\n',
 'centroidfile 1\n',
 'izernike 0 0 -4.0442156930541306e-05 \n',
 'izernike 0 1 -4.332919031597056e-05 \n',
 'izernike 0 2 9.573003193790922e-05 \n',
 'izernike 0 3 -0.0003705211361010626 \n',
 'izernike 0 4 -3.1817348860741566e-09 \n',
 'izernike 0 5 3.026344163991198e-08 \n',
 'izernike 0 6 2.004567507254775e-05 \n',
 'izernike 0 7 -9.10786523552383e-06 \n',
 'izernike 0 8 -1.4561234478195143e-08 \n',
 'izernike 0 9 -1.0309865121152308e-08 \n',
 'izernike 0 10 -1.3645986189662889e-05 \n',
 'izernike 0 11 7.554462746840323e-09 \n',
 'izernike 0 12 -1.65323745375795e-09 \n',
 'izernike 0 13 -3.6539217577959214e-08 \n',
 'izernike 0 14 2.4496827297845056e-10 \n',
 'izernike 0 15 1.4056091394706087e-06 \n',
 'izernike 0 16 -3.1352491529807935e-06 \n',
 'iz

Update `Opsim_obshistid`, which now will contain info about the particular combination of instrument, field, position, cmd file, resulting in a distinct exposure id for each simulated dataset. 

In [31]:
instDict =  {"comCam": 0, "lsstCam": 1}
invDict = {}
for key, value in zip(instDict.keys(), instDict.values()):
    print(key,value)
    invDict[value]=key

comCam 0
lsstCam 1


In [32]:
invDict

{0: 'comCam', 1: 'lsstCam'}

In [36]:
get_inst_dict()

{'comCam': 0, 'lsstCam': 1}

In [35]:
invert_dict(get_inst_dict())

{0: 'comCam', 1: 'lsstCam'}

In [44]:
def invert_dict(dic):
    invDic={}
    for key,value in zip(dic.keys(), dic.values()):
        invDic[value]=key
    return invDic

def get_inst_dict():
    return {"comCam": 0, "lsstCam": 1}

def get_field_dict():
    return {"high": 0, "med": 1, "low": 2, "Baade": 3}

def get_position_dict():
    return {"focal": 0, "extra": 1, "intra": 2}

def get_cmd_dict():
    return {
        "noBkgndPert00": 0,
        "noBkgndPert05": 1,
        "qckBkgndPert00": 2,
        "qckBkgndPert05": 3,
        "noBkgnd": 4,
        "qckBkgnd": 5,
    }

def calculate_obshistid(instrument, field, position, cmd_file, run):

    instDict = get_inst_dict()
    fieldDict = get_field_dict()
    positionDict = get_position_dict()

    if cmd_file.find("_") > 0:
        # eg 'noBkgndPert00_NCSA.cmd', 'noBkgndPert00_hyak.cmd'
        # i.e. with corrected surfacemap paths
        cmd = cmd_file.split("_")[0]
    else:  # eg. 'noBkgndPert00.cmd'
        # i.e. with original paths
        cmd = cmd_file.split(".")[0]
    cmdDict = get_cmd_dict()
    
    first = instDict[instrument]
    second = fieldDict[field]
    third = positionDict[position]
    fourth = cmdDict[cmd]
    obshistid = f"90{first}{second}{third}{fourth}{run}"
    return obshistid


def invert_obshistid(obshistid):
    ''' Given obshistid, invert the logic and find the 
    instrument, field, position, cmd_file, run
    '''
    instDictInv = invert_dict(get_inst_dict())
    fieldDictInv = invert_dict(get_field_dict())
    positionDictInv = invert_dict(get_position_dict())
    cmdDictInv = invert_dict(get_cmd_dict())
    
   # take last five digits
    digits = obshistid[-5:]

    first=int(digits[0])
    second=int(digits[1])
    third=int(digits[2])
    fourth=int(digits[3])
    run=int(digits[4])


    instrument = instDictInv[first]
    field = fieldDictInv[second]
    position= positionDictInv[third]
    cmd = cmdDictInv[fourth]
    
    return {'instrument':instrument, 'field':field, 
            'position':position, 'cmd':cmd, 
            'run':run
           }

In [24]:
digits

'00211'

In [60]:
import os
import numpy as np
from astropy.io import fits
run=1
for instrument in  ['comCam']:
    for field in ['high','med','low']:
        for pert in ['00','05']:
            for position in ['focal', 'extra', 'intra']:
            
                cmd_file = f'noBkgndPert{pert}.cmd'
                obshistid = calculate_obshistid(instrument,field,position,cmd_file,run)
                print(obshistid, ':', instrument, field, position, cmd_file, )
                #dic = invert_obshistid(obshistid)
                #print(dic)
                exposure = obshistid
                path = os.path.join(f'/project/scichris/aos/AOS/DM-28360\
/{instrument}/{field}/{position}/{obshistid}/repackaged/CC_H_20211231_{obshistid[-6:]}_R22_S00.fits')
                print(path)
                hdul = fits.open(path)
                seeing = []
                for j in range(7):
                    see = hdul[1].header[f'SEE{j}']
                    seeing.append(see)
                seeing_total = np.sqrt(np.sum(np.square(seeing)))
                seeing_fwhm = 2.355*seeing_total
                print(seeing_fwhm)
                               
                               
               

9000001 : comCam high focal noBkgndPert00.cmd
/project/scichris/aos/AOS/DM-28360/comCam/high/focal/9000001/repackaged/CC_H_20211231_000001_R22_S00.fits
0.5752073903703584
9000101 : comCam high extra noBkgndPert00.cmd
/project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000101/repackaged/CC_H_20211231_000101_R22_S00.fits
0.5752073903703584
9000201 : comCam high intra noBkgndPert00.cmd
/project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000201/repackaged/CC_H_20211231_000201_R22_S00.fits
0.5752073903703584
9000011 : comCam high focal noBkgndPert05.cmd
/project/scichris/aos/AOS/DM-28360/comCam/high/focal/9000011/repackaged/CC_H_20211231_000011_R22_S00.fits
0.5752073903703584
9000111 : comCam high extra noBkgndPert05.cmd
/project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000111/repackaged/CC_H_20211231_000111_R22_S00.fits
0.5752073903703584
9000211 : comCam high intra noBkgndPert05.cmd
/project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000211/repackaged/CC_H_20211231_000211_R22_S0

In [62]:
run=1
for instrument in  ['lsstCam']:
    for field in ['med','low']:
        for pert in ['00',]:
            for position in ['focal',]:
            
                cmd_file = f'qckBkgndPert{pert}.cmd'
                obshistid = calculate_obshistid(instrument,field,position,cmd_file,run)
                print(obshistid, ':', instrument, field, position, cmd_file, )

9011021 : lsstCam med focal qckBkgndPert00.cmd
9012021 : lsstCam low focal qckBkgndPert00.cmd


In [ ]:
Run these two just for corner 

In [16]:

def get_dirs_obshistids(instruments, fields, positions, cmd_files, root_dir, run):
    obshistids=[]
    work_dirs=[]
    for instrument in instruments:
        for field in fields:
            for position in positions:
                for cmd_file in cmd_files:
                    
                    # build the name of input directory with
                    # repackaged simulated files
                    # work_dir = os.path.join(root_dir, instrument, field, position)

                    obshistid = calculate_obshistid(
                        instrument, field, position, cmd_file, run
                    )

                    work_dir = os.path.join(
                        root_dir, instrument, field, position, str(obshistid)
                    )
                    
                    print(f"\n {instrument} {field} {position} {cmd_file} is {obshistid} and {work_dir}")
                    
                    obshistids.append(obshistid)
                    work_dirs.append(work_dir)
    return work_dirs, obshistids


In [52]:
def get_work_dirs_from_obshistids(obshistids, root_dir):
    work_dirs = []
    for obshistid in obshistids:
        dic = invert_obshistid(obshistid)
        work_dir = os.path.join(
                        root_dir, dic['instrument'], dic['field'], 
                        dic['position'], str(obshistid)
                    )
        work_dirs.append(work_dir)
    return work_dirs

In [17]:
instruments=['comCam']
fields=['high']
positions=['extra','intra', 'focal']
cmd_files =["noBkgndPert00_NCSA.cmd",  "noBkgndPert05_NCSA.cmd",
                 "qckBkgndPert00_NCSA.cmd", "qckBkgndPert05_NCSA.cmd"
                ]
root_dir = "/project/scichris/aos/AOS/DM-28360/"
run=1

In [53]:
work_dirs, obshistids = get_dirs_obshistids(instruments, fields, 
                                                    positions, cmd_files, root_dir, run)


 comCam high extra noBkgndPert00_NCSA.cmd is 9000101 and /project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000101

 comCam high extra noBkgndPert05_NCSA.cmd is 9000111 and /project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000111

 comCam high extra qckBkgndPert00_NCSA.cmd is 9000121 and /project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000121

 comCam high extra qckBkgndPert05_NCSA.cmd is 9000131 and /project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000131

 comCam high intra noBkgndPert00_NCSA.cmd is 9000201 and /project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000201

 comCam high intra noBkgndPert05_NCSA.cmd is 9000211 and /project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000211

 comCam high intra qckBkgndPert00_NCSA.cmd is 9000221 and /project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000221

 comCam high intra qckBkgndPert05_NCSA.cmd is 9000231 and /project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000231

 comCam high focal noBkgndPert00_NCSA.cmd i

In [54]:
work_dirs = get_work_dirs_from_obshistids(obshistids, root_dir)


In [55]:
for work_dir, obshistid in zip(work_dirs,obshistids):
        print( work_dir, obshistid)

/project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000101 9000101
/project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000111 9000111
/project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000121 9000121
/project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000131 9000131
/project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000201 9000201
/project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000211 9000211
/project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000221 9000221
/project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000231 9000231
/project/scichris/aos/AOS/DM-28360/comCam/high/focal/9000001 9000001
/project/scichris/aos/AOS/DM-28360/comCam/high/focal/9000011 9000011
/project/scichris/aos/AOS/DM-28360/comCam/high/focal/9000021 9000021
/project/scichris/aos/AOS/DM-28360/comCam/high/focal/9000031 9000031


comCam high obshistids to ingest and ISR : 

    python run_ps1_3_butler.py  -i  9000001  9000031  9000131  9000221 9000231  9000011  9000111  9000211  
    
    
but they all failed  with ISR on one chip - need to rerun with 

    python run_ps1_3_butler.py  -i 9000001  9000031  9000131  9000221 9000231  9000011  9000111  9000211   --clobber
    
    
    python run_ps1_3_butler.py  -i 9000031  --clobber    # worked fine 

In [ ]:
path = 
hdul = fits.open('')

In [13]:
inst_file= f'stars_{instrument}_PS1_{field}_{position}.inst'
path_to_inst_file = os.path.join(root_dir, inst_file)
header, content = readFile(path_to_inst_file)


In [20]:
content[0].find('obshistid')

6

In [67]:
instrument = 'comCam'
field = 'high'
position = 'focal'


def update_inst_file_obshistid(obshistid, inst_file,  root_dir, seeing=0.69):
    
    path_to_inst_file = os.path.join(root_dir, inst_file)

    header, content = readFile(path_to_inst_file)
    new_content = content.copy()
    for i in range(len(content)):
        # first update obshistid
        if content[i].find('obshistid')>0:
            split_line = content[i].split(' ')
            split_line[1] = f'{obshistid}\n'
            new_content[i] = ' '.join(split_line)
            
        # second update the seeing 
        
        if content[i].find('rawseeing')>0:
            split_line = content[i].split(' ')
            split_line[1] = f'{seeing}\n'
            new_content[i] = ' '.join(split_line)
            
            continue
            
    new_inst_file = inst_file[:-(len('.inst'))]+f'_{obshistid}.inst'
    out_file = os.path.join(root_dir, new_inst_file)
    write_to_file(out_file, new_content)
    print(f'Updated {inst_file} to {new_inst_file} in {root_dir}')


In [5]:
import os
root_dir = '/project/scichris/aos/AOS/DM-28360'
run=1
seeing=0.69
for instrument in  ['comCam']:
    for field in ['Baade',]:
        for position in ['focal', 'extra', 'intra']:
            for pert in ['00','05']:
                for bkgnd in ['qck','no']:
                    cmd_file = f'{bkgnd}BkgndPert{pert}_NCSA.cmd'
                    obshistid = calculate_obshistid(instrument,field,position,cmd_file,run)

                    inst_file= f'stars_{instrument}_PS1_{field}_{position}.inst'
                    update_inst_file_obshistid(obshistid, inst_file, root_dir, seeing ) 

Updated stars_comCam_PS1_Baade_focal.inst to stars_comCam_PS1_Baade_focal_9003021.inst in /project/scichris/aos/AOS/DM-28360
Updated stars_comCam_PS1_Baade_focal.inst to stars_comCam_PS1_Baade_focal_9003001.inst in /project/scichris/aos/AOS/DM-28360
Updated stars_comCam_PS1_Baade_focal.inst to stars_comCam_PS1_Baade_focal_9003031.inst in /project/scichris/aos/AOS/DM-28360
Updated stars_comCam_PS1_Baade_focal.inst to stars_comCam_PS1_Baade_focal_9003011.inst in /project/scichris/aos/AOS/DM-28360
Updated stars_comCam_PS1_Baade_extra.inst to stars_comCam_PS1_Baade_extra_9003121.inst in /project/scichris/aos/AOS/DM-28360
Updated stars_comCam_PS1_Baade_extra.inst to stars_comCam_PS1_Baade_extra_9003101.inst in /project/scichris/aos/AOS/DM-28360
Updated stars_comCam_PS1_Baade_extra.inst to stars_comCam_PS1_Baade_extra_9003131.inst in /project/scichris/aos/AOS/DM-28360
Updated stars_comCam_PS1_Baade_extra.inst to stars_comCam_PS1_Baade_extra_9003111.inst in /project/scichris/aos/AOS/DM-28360


In [68]:
import os
root_dir = '/project/scichris/aos/AOS/DM-28360'
run=1
for instrument in  ['lsstCam']:
    for field in ['low',]:
        for position in ['focal', 'extra', 'intra']:
            for pert in ['00','05']:
                cmd_file = f'qckBkgndPert{pert}.cmd'
                obshistid = calculate_obshistid(instrument,field,position,cmd_file,run)
                
                inst_file= f'stars_{instrument}_PS1_{field}_{position}.inst'
                update_inst_file_obshistid(obshistid, inst_file, root_dir ) 

Updated stars_lsstCam_PS1_low_focal.inst to stars_lsstCam_PS1_low_focal_9012021.inst in /project/scichris/aos/AOS/DM-28360
Updated stars_lsstCam_PS1_low_focal.inst to stars_lsstCam_PS1_low_focal_9012031.inst in /project/scichris/aos/AOS/DM-28360


FileNotFoundError: [Errno 2] No such file or directory: '/project/scichris/aos/AOS/DM-28360/stars_lsstCam_PS1_low_extra.inst'

In [63]:
run=1
for instrument in  ['lsstCam']:
    for field in ['high', 'med', 'low',]:
        for position in ['focal', 'extra', 'intra']:
            for pert in ['00','05']:
                cmd_file = f'qckBkgndPert{pert}.cmd'
                obshistid = calculate_obshistid(instrument,field,position,cmd_file,run)
                
                inst_file= f'stars_{instrument}_PS1_{field}_{position}.inst'
                update_inst_file_obshistid(obshistid, inst_file, root_dir ) 

NameError: name 'update_inst_file_obshistid' is not defined

In [107]:
work_dir = os.path.join(root_dir, instrument, field, position, str(obshistid))

In [6]:
cmd_file = 'noBkgndPert00.cmd'
cmd_file.find('_')
cmd_file.split('.')[0]

'noBkgndPert00'

In [4]:
cmd_file = 'noBkgndPert00_NCSA.cmd'
cmd_file.split('_')[0]

'noBkgndPert00'

In [2]:
instrument='comCam'
field='high'
position='intra'
cmd_file='qckBkgndPert00_NCSA.cmd'
run=1
calculate_obshistid(instrument, field, position, cmd_file, run)

'9000221'

python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions focal intra extra --cmd_file qckBkgndPert00.cmd 

    # default  
    # --instruments  comcam 
    # --fields high 
    # --positions focal 
    # --phosim_p  25
    # --cmd_file noBkgnd.cmd
    # --phosim_path /project/scichris/aos/phosim_syseng4/phosim.py
    # --root_dir /project/scichris/aos/ps1_phosim/



need to : 

    ssh scichris@lsst-login01.ncsa.illinois.edu 
    ssh lsst-devl03 
    source "/software/lsstsw/stack/loadLSST.bash"    # need that for python 
    setup lsst_distrib 
    
    source /project/scichris/aos/setup_aos_current.sh
    setup -k -r /project/scichris/aos/phosim_utils/
    

@@@ all below are    comcam  high  @@@ 

DONE  comHi1:  run on lsst-devl03  screen "pert" 
    python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions focal --phosim_p 24 --cmd_file noBkgndPert00_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/    
  ( output in /project/scichris/aos/AOS/DM-28360/comCam/high/focal/9000001/raw )  

DONE comHi3:  slurm 51781 from lsst-devl03 
    python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions extra --phosim_p 24  --cmd_file noBkgndPert00_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/
     ( /project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000101/raw )
     
     
DONE comHi5:  slurm 51782 from lsst-devl03 
    python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions intra --phosim_p 24 --cmd_file noBkgndPert00_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/
    (output in /project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000201/raw ) 
    
    
    
DONE comHi2:    slurm 51780 from lsst-devl03 
    python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions focal  --phosim_p 24 --cmd_file qckBkgndPert00_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/
   ( output in /project/scichris/aos/AOS/DM-28360/comCam/high/focal/9000021/raw )
   

    
DONE comHi6: slurm 51784 from lsst-devl03 
    python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions extra --phosim_p 24 --cmd_file qckBkgndPert00_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/
    ( output in /project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000121/raw ) 
    
    
@@@@@@@
NB: there are wrong paths in qckBkgndPert05.cmd ,  noBkgndPert05.cmd --> NEED to use the updated paths files !! 
@@@@@

comHi4:  ## running on screen "pert"  in lsst-devl02 
    python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions extra --phosim_p 24 --cmd_file qckBkgndPert00.cmd --root_dir /project/scichris/aos/AOS/DM-28360/
    (ctrl+a , d to detach  without killing the screen ...  ) 
    ( output in   
    
    
    python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions intra --phosim_p 24 --cmd_file qckBkgndPert00_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/
    (output in /project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000221/raw  )
    
(these were submitted from lsst-devl02 , ensuring that the stack and aos suite were sourced) 





comHi7:  slurm 51962 from lsst-devl03  (failed b/c there are wrong paths  in qckBkgnd ) 
         running slurm  51845
    python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions focal --phosim_p 24 --cmd_file qckBkgndPert05_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/  
    (output in   /project/scichris/aos/AOS/DM-28360/comCam/high/focal/9000031/raw ) 
 
comHi8: slurm 51964
     python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions intra --phosim_p 24 --cmd_file qckBkgndPert05_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/ 
    (output /project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000231/raw ) 
    
comHi9:  slurm51963
     python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions extra --phosim_p 24 --cmd_file qckBkgndPert05_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/ 
     ( output in /project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000131/raw )



DONE...

comHi10:  slurm51965
    python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions focal --phosim_p 24 --cmd_file noBkgndPert05_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/  
     ( output  /project/scichris/aos/AOS/DM-28360/comCam/high/focal/9000011/raw ) 


comHi11:  slurm51967
    python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions intra --phosim_p 24 --cmd_file noBkgndPert05_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/ 
     ( output in /project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000211/raw  ) 

comHi12:  slurm51966
    python /project/scichris/aos/AOS/run_ps1_phosim.py  --positions extra --phosim_p 24 --cmd_file noBkgndPert05_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/  
    ( ouptut uin  /project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000111/raw ) 

    

Also submitted a 10-node request to slurm - if that number were available, it'd be done very soon (~`1 day).


    python /gscratch/astro/suberlak/aos/AOS/run_ps1_phosim.py --instruments lsstCam --fields high --positions extra --root_dir /gscratch/astro/suberlak/aos/AOS/DM-28360 --cmd_file qckBkgndPert00_hyak.cmd --phosim_path /gscratch/astro/brycek/ts_repos/phosim_syseng4/phosim.py --phosim_p 192

(cancelled ... ) 

in runSlurmLsst.sl 




To ingest and run ISR on those that are finished (making a repo inside each directory)

In [ ]:
python /home/scichris/notebooks/run_ps1_3_butler.py --instruments comCam --fields high --positions 

Total : 12 above, i.e.  comcam x high x {intra,extra,focal} x {noBkgnd, qckBkgnd} x {smallPert, largePert}
    
    

In [6]:
import os
import subprocess
#Couldn't I have a script to make slurm scripts requesting one-node jobs ? 

def write_to_file(out_file, content):
    with open(out_file, "w") as output:
        for line in content:
            output.write(line)
            

def write_slurm_script(python_script = '/project/scichris/aos/AOS/run_ps1_phosim.py',
                   instrument = 'comCam',
                   field = 'high',
                   position = 'extra',
                   phosim_p = 24,
                   cmd_file = "noBkgndPert05_NCSA.cmd",
                   root_dir = '/project/scichris/aos/AOS/DM-28360/',
                   nodes = 1, ntasks=24, job_name="comHi12",
                   slurm_file_name = "runSlurmTest.sl"
                  ):

    cmd = f"python {python_script} --instruments {instrument} --fields {field} \
    --positions {position} --phosim_p {phosim_p} --cmd_file {cmd_file} --root_dir {root_dir} \n"
    
    content = ["#!/bin/bash -l \n",
               "#SBATCH --partition normal \n"
                f"#SBATCH --nodes {nodes} \n",
                f"#SBATCH --ntasks {ntasks} \n",
                "#SBATCH -t 500:00:00 \n",
                f"#SBATCH --job-name {job_name} \n",
                'echo "starting at `date` on `hostname`" \n',
                'pwd \n',
                 cmd,
                'echo "ended at `date` on `hostname`" \n']
    out_file = f'/project/scichris/aos/AOS/{slurm_file_name}'
    write_to_file(out_file, content)
    return out_file
    
    
#slurm_file = write_slurm_script()


# also, a script to submit the one-job slurm script automatically...
def submit_slurm_job(slurm_file):
    print(f'Running sbatch  {slurm_file}')
    subprocess.call(['sbatch', slurm_file]) 



I pasted the below to `run_ps1_slurm.py` ,  I could add `argparse` to slurm to make it even more slick, but
it makes it easier for now as a way to submit lots of N-node jobs...


I pasted the cell below  to `run_ps1_slurm.py` and executed from `lsst-devl02` to submit a suite of 1-node jobs to slurm, counting from 7 to 12...  Made sure I'm in a screen where `setup_aos_current.sh` had been sourced, and I set `phosim_utils` to current master (so that `--focus z` arg is available and `ROTANG` is also written in a correct fashion)... 

In [11]:
instrument='comCam'
field='high'
counter = 7
run=1
for bkgnd in ['qck','no']:
    cmd_file = f'{bkgnd}BkgndPert05_NCSA.cmd' 
    for position in ['focal','extra','intra']:
        job_name=f"comHi{counter}"
        slurm_file_name = f"runSlurm{job_name}.sl"
        
        obshistid = calculate_obshistid(instrument,field,position,cmd_file,run)
        print(instrument, field, position, cmd_file, obshistid)
#         slurm_file = write_slurm_script(position=position, 
#                                         cmd_file=cmd_file, 
#                                         job_name=job_name,
#                                         slurm_file_name=slurm_file_name
#                                        )
        
#         submit_slurm_job(slurm_file)
        

comCam high focal qckBkgndPert05_NCSA.cmd 9000031
comCam high extra qckBkgndPert05_NCSA.cmd 9000131
comCam high intra qckBkgndPert05_NCSA.cmd 9000231
comCam high focal noBkgndPert05_NCSA.cmd 9000011
comCam high extra noBkgndPert05_NCSA.cmd 9000111
comCam high intra noBkgndPert05_NCSA.cmd 9000211


## Check the airmass and seeing for those simulations.. 

In [1]:
from astropy.io import fits
import os 

files = ['/project/scichris/aos/AOS/DM-28360/comCam/high/intra/9000211/repackaged/CC_H_20211231_000211_R22_S00.fits',
        '/project/scichris/aos/AOS/DM-28360/comCam/high/extra/9000101/repackaged/CC_H_20211231_000101_R22_S00.fits' 
        ]
for file in files:
    hdul  = fits.open(file)
    print(hdul[0].header['AMSTART'])

1.24813143788881
1.24813143788881


In [22]:
import numpy as np
s = []
for i in range(6):
    s.append(hdul[1].header[f'SEE{i}'])
np.mean(s)

0.046560500000000005

In [24]:
np.sum(s)

0.27936300000000003

The median at-zenith at 500nm von Karman seeing for the LSST site is 0.69 arcseconds. (L. Jones) 

Ok, since this file https://github.com/lsst-ts/phosim_syseng4/blob/6105fcf9d65b37d2d157be1f99d70dc263da0e76/validation/speed_trim.pro uses `Opsim_rawseeing` as `Opsim_rawseeing 0.65`, I think it was supposed to be  the LSST value, since it is very close to what  Lynne gave me. 

Also, the LSST sensitivity matrix has the same https://github.com/lsst-ts/phosim_syseng4/blob/6105fcf9d65b37d2d157be1f99d70dc263da0e76/data/lsst_sen_matrix/sensitivity_matrix.inst `Opsim_rawseeing 0.67`

Which seeing to use?   There are 7 atmosphere files... 
But I see that the inst files pass `Opsim_rawseeing -1` ,  and `-1` means that seeing is chosen at random (according to reference file). 

Here is where `phosim.py` reads the `seeing` from instance catalog: https://github.com/lsst-ts/phosim_syseng4/blob/6105fcf9d65b37d2d157be1f99d70dc263da0e76/phosim.py#L221


          elif "seeing" in line:
          self.constrainseeing=lstr[1]
          
which sets the values of the `constrainseeing` variable , saved in `obs_'+self.observationID+'.pars` file 

And this file https://github.com/lsst-ts/phosim_syseng4/blob/6105fcf9d65b37d2d157be1f99d70dc263da0e76/source/atmosphere/atmosphere_creator.cpp#L64 

points me to  a paper "Atmospheric Parameters for the Image
    /// Simulator Based on the Site of the Large Synoptic Survey Telescope" by Young 
 but I can't find it anywhere...
 
 
 But main Petersons paper  https://iopscience.iop.org/article/10.1088/0067-0049/218/1/14/pdf  has in section 1.4 a description of LSST seeing, and indeed it assumes it is 0.67 arcsec.  
 
 So in other words, `Opsim_rawseeing`  is in units of arcsec, and I should set it to 
 
     Opsim_rawseeing 0.69
    


In [7]:
import numpy as np 
0.210822*(2.0*np.sqrt(2.0*np.log(2.0)))

0.49644787153351483

In [ ]:
ATMFILE =                    7 / Number of atmosphere files                     
AFILE0  = 'atmospherescreen_9000101_0' / Atmosphere file                        
CFILE0  = '        '           / Cloud file                                     
SEE0    =             0.070356 / Seeing at 5000 angstrom (sigma)                
WIND0   =             4.458112 / Wind speed (m/s)                               
WDIR0   =           234.190548 / Wind direction (degrees)                       
OSCL0   =             24.03503 / Outer scale (m)                                
HGT0    =                  16. / Height (km)                                    
CMEAN0  =                   0. / Mean cloud extinction (mag)                    
CVARY0  =                   0. / Variation of cloud ext. (mag)                  
AFILE1  = 'atmospherescreen_9000101_1' / Atmosphere file                        
CFILE1  = 'cloudscreen_9000101_1' / Cloud file                                  
SEE1    =             0.060413 / Seeing at 5000 angstrom (sigma)                
WIND1   =            26.849277 / Wind speed (m/s)                               
WDIR1   =           351.085985 / Wind direction (degrees)                       
OSCL1   =            52.079948 / Outer scale (m)                                
HGT1    =                   8. / Height (km)                                    
CMEAN1  =                   0. / Mean cloud extinction (mag)                    
CVARY1  =                   0. / Variation of cloud ext. (mag)                  
AFILE2  = 'atmospherescreen_9000101_2' / Atmosphere file                        
CFILE2  = 'cloudscreen_9000101_2' / Cloud file                                  
SEE2    =             0.046016 / Seeing at 5000 angstrom (sigma)                
WIND2   =            13.627671 / Wind speed (m/s)                               
WDIR2   =           353.365192 / Wind direction (degrees)                       
OSCL2   =            31.267035 / Outer scale (m)                                
HGT2    =                   4. / Height (km)                                    
CMEAN2  =                   0. / Mean cloud extinction (mag)                    
CVARY2  =                   0. / Variation of cloud ext. (mag)                  
AFILE3  = 'atmospherescreen_9000101_3' / Atmosphere file                        
CFILE3  = '        '           / Cloud file                                     
SEE3    =             0.008804 / Seeing at 5000 angstrom (sigma)                
WIND3   =             7.611568 / Wind speed (m/s)                               
WDIR3   =           341.760454 / Wind direction (degrees)                       
OSCL3   =            25.557991 / Outer scale (m)                                
HGT3    =                   2. / Height (km)                                    
CMEAN3  =                   0. / Mean cloud extinction (mag)                    
CVARY3  =                   0. / Variation of cloud ext. (mag)                  
AFILE4  = 'atmospherescreen_9000101_4' / Atmosphere file                        
CFILE4  = '        '           / Cloud file                                     
SEE4    =             0.043693 / Seeing at 5000 angstrom (sigma)                
WIND4   =             5.123655 / Wind speed (m/s)                               
WDIR4   =           334.082393 / Wind direction (degrees)                       
OSCL4   =             56.96442 / Outer scale (m)                                
HGT4    =                   1. / Height (km)                                    
CMEAN4  =                   0. / Mean cloud extinction (mag)                    
CVARY4  =                   0. / Variation of cloud ext. (mag)                  
AFILE5  = 'atmospherescreen_9000101_5' / Atmosphere file                        
CFILE5  = '        '           / Cloud file                                     
SEE5    =             0.050081 / Seeing at 5000 angstrom (sigma)                
WIND5   =             3.964125 / Wind speed (m/s)                               
WDIR5   =           331.093557 / Wind direction (degrees)                       
OSCL5   =            20.492025 / Outer scale (m)                                
HGT5    =                  0.5 / Height (km)                                    
CMEAN5  =                   0. / Mean cloud extinction (mag)                    
CVARY5  =                   0. / Variation of cloud ext. (mag)                  
AFILE6  = 'atmospherescreen_9000101_6' / Atmosphere file                        
CFILE6  = '        '           / Cloud file                                     
SEE6    =             0.210822 / Seeing at 5000 angstrom (sigma)                
WIND6   =             3.197529 / Wind speed (m/s)                               
WDIR6   =            324.49778 / Wind direction (degrees)                       
OSCL6   =                  10. / Outer scale (m)                                
HGT6    =                 0.02 / Height (km)                                    
CMEAN6  =                   0. / Mean cloud extinction (mag)                    
CVARY6  =                   0. / Variation of cloud ext. (mag)                  
RAYGRAD =             0.000712 / Density gradient                               
O2GRAD  =              3.8E-05 / O2 gradient                                    
O3GRAD  =             0.034217 / O3 gradient                                    
H2OGRAD =             0.049043 / H2O gradient                                   
AERGRAD =             0.001176 / Aerosol gradient                               
RAYANG  =             0.438685 / Density angle                                  
O2ANG   =             4.415017 / O2 angle                                       
O3ANG   =             4.605518 / O3 angle                                       
H2OANG  =              5.81092 / H2O angle                                      
AERANG  =             5.393528 / Aerosol angle                                  
ATMDISP =                    1 / Atmos. dispersion (0=off/1=on)                 
PRESS   =                 520. / Air pressure (mmHg)                            
H2OPRESS=                   8. / Water vapor pressure (mmHg)                    
TEMPERA =                  20. / Ground temperature (degrees C)                 
RELDENS =             0.997984 / Relative density                               
RELO2   =             0.999896 / Relative O2 fraction                           
RELH2O  =             1.150661 / Relative H2O fraction                          
AERTAU  =              0.01506 / Aerosol optical depth                          
AERIND  =                -1.28 / Aerosol index                                  
RELO3   =             0.926995 / Relative O3 fraction                           
LASCPR  =                   0. / Large angle scattering probability             
DOMESEE =                  0.1 / Dome Seeing (arcseconds FWHM)                  
TOYPSF  =                   0. / Toy PSF (arcseconds FWHM) 

In [ ]:
This is the outcome of 

    seefactor[i]/(pow(1/cos(zenith), 0.6))

in https://github.com/lsst-ts/phosim_syseng4/blob/6105fcf9d65b37d2d157be1f99d70dc263da0e76/source/raytrace/settings.cpp 

# Running comCam medium on NCSA, submitting various combinations with updated seeing (0.69)

on lsst-devl02,  in  a screen :
    
    python run_ps1_slurm.py --instruments comCam --fields med --suffix NCSA --job_prefix comMd --dry_run 
    
    ( by default, other params are 
     --python_script   /project/scichris/aos/AOS/run_ps1_phosim.py
     --positions focal extra intra
     --backgrounds qck no 
     --perts 00 05 
     --phosim_p 24
     --root_dir /project/scichris/aos/AOS/DM-28360/
     --nodes 1
     --ntasks 24
     --time_limit 300
     --slurm_file runSlurm
     --slurm_path /project/scichris/aos/AOS/  
    )

# Running comCam low  on NCSA, submitting various combos with updated seeing (0.69):

I first run the `update_inst_file_obshistid` with the `low` field. With 2 node-hours per CCD for in-focus low-latitude image, and three times as much for defocal image, I estimate that the single combo of in-focal image would take 2*9 CCDs = 18 node-hours, and defocal image would take 3*2*9 CCDs = 54 node-hours.  Thus it seems best to ask for 2 nodes for each job, so it's completed in under a day (<~9 and 27 hrs, respectively). I submit slurm jobs: 


    ssh lsst-devl02
    scren -x pert (I already setup the repackager there .)
    
    cd /project/scichris/aos/AOS
    git pull    ( to make sure I'm using updated runSlurm, etc. ) 
    
    
    python run_ps1_slurm.py --instruments comCam --fields low --suffix NCSA --job_prefix comLo --nodes 2 --ntasks 48 --phosim_p 48 --dry_run 
    
    
( by default, other params are 
 --python_script   /project/scichris/aos/AOS/run_ps1_phosim.py
 --positions focal extra intra
 --backgrounds qck no 
 --perts 00 05 
 --root_dir /project/scichris/aos/AOS/DM-28360/
 --time_limit 300
 --slurm_file runSlurm
 --slurm_path /project/scichris/aos/AOS/  ) 

## Running comCam Baade (NCSA slurm) 

 ssh lsst-devl02
    scren -x pert (I already setup the repackager there .)
    cd /project/scichris/aos/AOS
    
    python run_ps1_slurm.py --fields Baade --nodes 2 --ntasks 48 --phosim_p 48 --job_prefix comBa

Doing the ingest and ISR: 

    python run_ps1_3_butler.py --fields Baade --positions intra extra focal --dry_run

# Ingest and ISR all medium and low comcam simulations - use a script 

I edit "run_ps1_3_butler.py", adding just checking for the dirs that are there under 'med' and 'low' instead of having to know...




In [2]:
os.listdir(work_dir)

['9001221', '9001201', '9001231', '9001211']

In [7]:
import os
def find_dirs(root_dir='/project/scichris/aos/AOS/DM-28360/', instruments=['comCam'], 
              fields=['med'], position=['focal', 'extra', 'intra']):

    all_dirs = []

    for instrument in instruments:
        for field in fields:
            for position in positions:
                work_dir = os.path.join(root_dir, instrument, field, position)
                print(work_dir)
                sub_dir_names = os.listdir(work_dir)
                for sub_dir in sub_dir_names:
                    all_dirs.append(os.path.join(work_dir,sub_dir))
                # add 
                print(' ', sub_dirs)

    return all_dirs

In [6]:
( by default,  
--instruments  ['comCam']
--root_dir '/project/scichris/aos/AOS/DM-28360/'
 
 

cd /home/scichris/notebooks/
python run_ps1_3_butler.py --fields med low --positions focal extra intra --dry_run

['/project/scichris/aos/AOS/DM-28360/comCam/med/focal/9001011',
 '/project/scichris/aos/AOS/DM-28360/comCam/med/focal/9001001',
 '/project/scichris/aos/AOS/DM-28360/comCam/med/focal/9001021',
 '/project/scichris/aos/AOS/DM-28360/comCam/med/focal/9001031',
 '/project/scichris/aos/AOS/DM-28360/comCam/med/extra/9001101',
 '/project/scichris/aos/AOS/DM-28360/comCam/med/extra/9001121',
 '/project/scichris/aos/AOS/DM-28360/comCam/med/extra/9001111',
 '/project/scichris/aos/AOS/DM-28360/comCam/med/extra/9001131',
 '/project/scichris/aos/AOS/DM-28360/comCam/med/intra/9001221',
 '/project/scichris/aos/AOS/DM-28360/comCam/med/intra/9001201',
 '/project/scichris/aos/AOS/DM-28360/comCam/med/intra/9001231',
 '/project/scichris/aos/AOS/DM-28360/comCam/med/intra/9001211']

# lsstCam : Running just `stars_lsstCam_PS1_high_intra.inst` on hyak with `qckBkgndPert00.cmd` - find out what should be the obshistid: 

In [3]:
run=1
for instrument in  ['lsstCam']:
    for field in ['high']:
        for position in ['extra',]:
            for pert in ['00',]:
                cmd_file = f'qckBkgndPert{pert}.cmd'
                obshistid = calculate_obshistid(instrument,field,position,cmd_file,run)
                print(obshistid, ':', instrument, field, position, cmd_file, )

9010121 : lsstCam high extra qckBkgndPert00.cmd


Manually change that `obshistid` in that file ... 

Update the path for running on  UW hyak :

    python run_ps1_fix_cmd_path.py --root_dir /gscratch/astro/suberlak/aos/AOS/DM-28360 --suffix hyak 

Run phosim with a slurm script 

Asking for 8 nodes, with 24 tasks per node, I could do -p of 8*24=192 

    source /gscratch/astro/suberlak/aos/setup_aos.sh 


    
    File: runSlurm.sl                                                                                                  

    #!/bin/bash -l
    #SBATCH --account astro
    #SBATCH --partition astro
    #SBATCH --nodes 8
    #SBATCH --ntasks-per-node 24
    #SBATCH -t 5000:00:00
    #SBATCH --job-name lsstHq00
    #SBATCH --mem=246G  # 256GB is the smallest astro node , so if that's the only available select 246G here
    # Specify the real memory required per node

    ##turn on e-mail notification
    #SBATCH --mail-type=ALL
    #SBATCH --mail-user=suberlak@uw.edu

    ## export all your environment variables to the batch job session
    #SBATCH --export=all


    ## may not need that one, since everywhere I provide abspath ...
    ## specify working directory for the  job ...

    #SBATCH --chdir=/gscratch/astro/suberlak/aos/
    echo "starting at `date` on `hostname`"

    pwd

    python /gscratch/astro/suberlak/aos/AOS/run_ps1_phosim.py --instruments lsstCam --fields high --positions extra --root_dir /gscratch/astro/suberlak/aos/AOS/DM-28360 --cmd_file qckBkgndPert00_hyak.cmd --phosim_path /gscratch/astro/brycek/ts_repos/phosim_syseng4/phosim.py --phosim_p 192

    echo "ended at `date` on `hostname`"



In [27]:
root_dir = "/project/scichris/aos/AOS/DM-28360"
pert='00'
surfacemap_dir = os.path.join(f"imgCloseLoop_0-{pert}", "iter0", "pert")
mirror_file = "M1res.txt"

path_to_mirror = os.path.join(root_dir, surfacemap_dir, mirror_file)
print(path_to_mirror)

/project/scichris/aos/AOS/DM-28360/imgCloseLoop_0-00/iter0/pert/M1res.txt


It stopped after 122 sensors because of timeout, so first find out what the sensors were:

    ssh mox 
    
    cd /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/
    
    # change the name of the  'raw' directory:
    mv raw raw_old
    
    ls raw_old/lsst_e_*  > list.txt
    
    

In [2]:
import numpy as np 
fpath = '/project/scichris/aos/AOS/DM-28360/list.txt'
files = np.loadtxt(fpath, dtype=str)
sensorsDone = ['_'.join(fileName.split('_')[-3:-1]) for fileName in files]
len(np.array(sensorsDone))

122

In [4]:
from lsst.obs.lsst import LsstCam
camera = LsstCam().getCamera()
sensorsAll = list(camera.getNameMap().keys())
sensorsToDo = np.array(sensorsAll)[~np.in1d(sensorsAll,sensorsDone)]
len(sensorsToDo)

83

In [13]:
len(sensorsAll) / 5

41.0

In [17]:
import run_ps1_functions as func
import imp
imp.reload(func)


<module 'run_ps1_functions' from '/home/scichris/notebooks/run_ps1_functions.py'>

In [19]:
sensors = func.get_camera_sensors('lsstCam')

In [20]:
sensors

array(['R00_SG0', 'R00_SG1', 'R00_SW0', 'R00_SW1', 'R01_S00', 'R01_S01',
       'R01_S02', 'R01_S10', 'R01_S11', 'R01_S12', 'R01_S20', 'R01_S21',
       'R01_S22', 'R02_S00', 'R02_S01', 'R02_S02', 'R02_S10', 'R02_S11',
       'R02_S12', 'R02_S20', 'R02_S21', 'R02_S22', 'R03_S00', 'R03_S01',
       'R03_S02', 'R03_S10', 'R03_S11', 'R03_S12', 'R03_S20', 'R03_S21',
       'R03_S22', 'R04_SG0', 'R04_SG1', 'R04_SW0', 'R04_SW1', 'R10_S00',
       'R10_S01', 'R10_S02', 'R10_S10', 'R10_S11', 'R10_S12', 'R10_S20',
       'R10_S21', 'R10_S22', 'R11_S00', 'R11_S01', 'R11_S02', 'R11_S10',
       'R11_S11', 'R11_S12', 'R11_S20', 'R11_S21', 'R11_S22', 'R12_S00',
       'R12_S01', 'R12_S02', 'R12_S10', 'R12_S11', 'R12_S12', 'R12_S20',
       'R12_S21', 'R12_S22', 'R13_S00', 'R13_S01', 'R13_S02', 'R13_S10',
       'R13_S11', 'R13_S12', 'R13_S20', 'R13_S21', 'R13_S22', 'R14_S00',
       'R14_S01', 'R14_S02', 'R14_S10', 'R14_S11', 'R14_S12', 'R14_S20',
       'R14_S21', 'R14_S22', 'R20_S00', 'R20_S01', 

In [23]:
sensors = np.sort(sensorsAll)
sensor_groups = np.array_split(sensors,5)

In [21]:
select = 5
select -= 1 
select

4

In [24]:
sensors_select = sensor_groups[0]
sensors_select

array(['R00_SG0', 'R00_SG1', 'R00_SW0', 'R00_SW1', 'R01_S00', 'R01_S01',
       'R01_S02', 'R01_S10', 'R01_S11', 'R01_S12', 'R01_S20', 'R01_S21',
       'R01_S22', 'R02_S00', 'R02_S01', 'R02_S02', 'R02_S10', 'R02_S11',
       'R02_S12', 'R02_S20', 'R02_S21', 'R02_S22', 'R03_S00', 'R03_S01',
       'R03_S02', 'R03_S10', 'R03_S11', 'R03_S12', 'R03_S20', 'R03_S21',
       'R03_S22', 'R04_SG0', 'R04_SG1', 'R04_SW0', 'R04_SW1', 'R10_S00',
       'R10_S01', 'R10_S02', 'R10_S10', 'R10_S11', 'R10_S12'], dtype='<U7')

In [25]:
sensor_names = func.sensor_list_to_string(sensors_select)

In [26]:
sensor_names

'R00_SG0|R00_SG1|R00_SW0|R00_SW1|R01_S00|R01_S01|R01_S02|R01_S10|R01_S11|R01_S12|R01_S20|R01_S21|R01_S22|R02_S00|R02_S01|R02_S02|R02_S10|R02_S11|R02_S12|R02_S20|R02_S21|R02_S22|R03_S00|R03_S01|R03_S02|R03_S10|R03_S11|R03_S12|R03_S20|R03_S21|R03_S22|R04_SG0|R04_SG1|R04_SW0|R04_SW1|R10_S00|R10_S01|R10_S02|R10_S10|R10_S11|R10_S12|'

In [ ]:
python

In [34]:
def sensor_list_to_string(sensorNameList):
    sensors = ""
    for sensor in sensorNameList:
        sensors += "%s|" % sensor
    return sensors

sensorsToDoStr = sensor_list_to_string(sensorsToDo)

In [35]:
sensorsToDoStr

'R44_SW1|R44_SW0|R40_SW1|R40_SW0|R41_S22|R20_S11|R14_S02|R43_S00|R20_S00|R22_S02|R12_S11|R44_SG0|R10_S20|R03_S02|R43_S12|R04_SG1|R33_S12|R34_S22|R32_S01|R41_S02|R32_S02|R01_S02|R34_S00|R03_S10|R41_S10|R01_S21|R33_S10|R34_S11|R43_S22|R24_S11|R42_S11|R00_SG1|R12_S20|R22_S21|R24_S22|R33_S11|R30_S10|R42_S10|R33_S02|R13_S12|R33_S21|R30_S21|R32_S21|R41_S20|R31_S10|R32_S22|R32_S10|R43_S11|R32_S11|R34_S10|R32_S12|R33_S01|R34_S21|R32_S20|R33_S00|R33_S20|R33_S22|R34_S01|R34_S02|R34_S12|R00_SG0|R34_S20|R41_S00|R41_S01|R41_S11|R42_S02|R41_S12|R41_S21|R42_S01|R42_S12|R42_S20|R42_S21|R42_S00|R42_S22|R43_S01|R43_S02|R43_S10|R44_SG1|R43_S20|R43_S21|R04_SG0|R40_SG0|R40_SG1|'

Make a sensor selection... Just submit the appropriate "run phosim" command, and store it manually (rather than using the `run_ps1_phosim.py` - can find out what command was used with `--dry_run` :

    python /gscratch/astro/suberlak/aos/AOS/run_ps1_phosim.py --instruments lsstCam --fields high --positions extra --root_dir /gscratch/astro/suberlak/aos/AOS/DM-28360 --cmd_file qckBkgndPert00_hyak.cmd --phosim_path /gscratch/astro/brycek/ts_repos/phosim_syseng4/phosim.py --phosim_p 192 --dry_run


yields 

    python /gscratch/astro/brycek/ts_repos/phosim_syseng4/phosim.py /gscratch/astro/suberlak/aos/AOS/DM-28360/stars_lsstCam_PS1_high_extra_9010121.inst -i lsst -e 1     -c /gscratch/astro/suberlak/aos/AOS/DM-28360/qckBkgndPert00_hyak.cmd -w /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/raw -p 192  -o /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/raw > /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/starPhoSim.log 2>&1
 
    phosim_repackager.py /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/raw     --out_dir /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/repackaged --inst lsst --focusz -1500
  
  

So I modify the phosim command, adding -s : 

    python /gscratch/astro/brycek/ts_repos/phosim_syseng4/phosim.py /gscratch/astro/suberlak/aos/AOS/DM-28360/stars_lsstCam_PS1_high_extra_9010121.inst -i lsst -e 1  -s  "R44_SW1|R44_SW0|R40_SW1|R40_SW0|R41_S22|R20_S11|R14_S02|R43_S00|R20_S00|R22_S02|R12_S11|R44_SG0|R10_S20|R03_S02|R43_S12|R04_SG1|R33_S12|R34_S22|R32_S01|R41_S02|R32_S02|R01_S02|R34_S00|R03_S10|R41_S10|R01_S21|R33_S10|R34_S11|R43_S22|R24_S11|R42_S11|R00_SG1|R12_S20|R22_S21|R24_S22|R33_S11|R30_S10|R42_S10|R33_S02|R13_S12|R33_S21|R30_S21|R32_S21|R41_S20|R31_S10|R32_S22|R32_S10|R43_S11|R32_S11|R34_S10|R32_S12|R33_S01|R34_S21|R32_S20|R33_S00|R33_S20|R33_S22|R34_S01|R34_S02|R34_S12|R00_SG0|R34_S20|R41_S00|R41_S01|R41_S11|R42_S02|R41_S12|R41_S21|R42_S01|R42_S12|R42_S20|R42_S21|R42_S00|R42_S22|R43_S01|R43_S02|R43_S10|R44_SG1|R43_S20|R43_S21|R04_SG0|R40_SG0|R40_SG1"   -c /gscratch/astro/suberlak/aos/AOS/DM-28360/qckBkgndPert00_hyak.cmd -w /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/raw -p 192  -o /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/raw > /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/starPhoSim.log 2>&1

and I modify the slurm script to just run that : 
    
    

File: runSlurmLsstCam.sl                                                                                                 

    #!/bin/bash -l
    #SBATCH --account astro
    #SBATCH --partition astro
    #SBATCH --nodes 8
    #SBATCH --ntasks-per-node 24
    #SBATCH -t 5000:00:00
    #SBATCH --job-name lsstHq00
    #SBATCH --mem=246G  # 256GB is the smallest astro node , so if that's the only available select 246G here
    # Specify the real memory required per node

    ##turn on e-mail notification
    #SBATCH --mail-type=ALL
    #SBATCH --mail-user=suberlak@uw.edu

    ## export all your environment variables to the batch job session
    #SBATCH --export=all


    ## may not need that one, since everywhere I provide abspath ...
    ## specify working directory for the  job ...

    #SBATCH --chdir=/gscratch/astro/suberlak/aos/
    echo "starting at `date` on `hostname`"

    pwd

    python /gscratch/astro/brycek/ts_repos/phosim_syseng4/phosim.py /gscratch/astro/suberlak/aos/AOS/DM-28360/stars_lsstCam_PS1_high_extra_9010121.inst -i lsst -e 1  -s  "R44_SW1|R44_SW0|R40_SW1|R40_SW0|R41_S22|R20_S11|R14_S02|R43_S00|R20_S00|R22_S02|R12_S11|R44_SG0|R10_S20|R03_S02|R43_S12|R04_SG1|R33_S12|R34_S22|R32_S01|R41_S02|R32_S02|R01_S02|R34_S00|R03_S10|R41_S10|R01_S21|R33_S10|R34_S11|R43_S22|R24_S11|R42_S11|R00_SG1|R12_S20|R22_S21|R24_S22|R33_S11|R30_S10|R42_S10|R33_S02|R13_S12|R33_S21|R30_S21|R32_S21|R41_S20|R31_S10|R32_S22|R32_S10|R43_S11|R32_S11|R34_S10|R32_S12|R33_S01|R34_S21|R32_S20|R33_S00|R33_S20|R33_S22|R34_S01|R34_S02|R34_S12|R00_SG0|R34_S20|R41_S00|R41_S01|R41_S11|R42_S02|R41_S12|R41_S21|R42_S01|R42_S12|R42_S20|R42_S21|R42_S00|R42_S22|R43_S01|R43_S02|R43_S10|R44_SG1|R43_S20|R43_S21|R04_SG0|R40_SG0|R40_SG1"   -c /gscratch/astro/suberlak/aos/AOS/DM-28360/qckBkgndPert00_hyak.cmd -w /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/raw -p 288  -o /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/raw > /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/starPhoSim2.log 2>&1


    echo "ended at `date` on `hostname`"



I submit on the screen where I enabled the stack and the AOS packages:
    
    screen -S aos 
    source /gscratch/astro/suberlak/aos/setup_aos.sh 
    
    

I checked the queue with 

    squeue -p astro
    
and the resources with 

    sinfo -p astro
    
    
and submitted with 
 
    sbatch runSlurmLsstCam.sl
    
    
But even with 1500 hrs it still claims its reserved for maintenance ... 

This is because there is a maintenance every first Tuesday of the month, the nearest one being on Tuesday June 9  - see https://wiki.cac.washington.edu/display/hyakusers/Maintenance+schedule 
    

I check with this website https://datetimego.com/how-many-hours-until-9am-tuesday  that it's about 88 hrs from now, so I set the max time to be 80 hrs. 

Since it takes 3 node-hours for high galactic latitude per CCD for defocal image, and we need  to finish 88 CCDs, i.e. need 3*88 node-hours = 264 node-hours, so   with 12 nodes we should be done in 264 node-hours / 12 nodes = 22 hrs , i.e. by next day....


I wonder if that's at all true ! 
    
    

Next : need to repackage ALL  - combine  /raw/ and /raw_old/  directories into /raw/ , and run 
    
    
    phosim_repackager.py /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/raw     --out_dir /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/repackaged --inst lsst --focusz -1500


But I have some b izarre segmentation fault error ... 


Try copying all the 3088 amp files to NCSA . On hyak , run 

    scp -rp /gscratch/astro/suberlak/aos/AOS/DM-28360/lsstCam/high/extra/9010121/raw/  scichris@lsst-login01.ncsa.illinois.edu:/project/scichris/aos/AOS/DM-28360/lsstCam/high/extra/9010121/raw/


And now on NCSA LSP, run 

    phosim_repackager.py /project/scichris/aos/AOS/DM-28360/lsstCam/high/extra/9010121/raw/raw     --out_dir /project/scichris/aos/AOS/DM-28360/lsstCam/high/extra/9010121/repackaged --inst lsst --focusz -1500

Then do the gen3 ingest and ISR:


    python run_ps1_3_butler.py --obshistids 9010121 (--dry_run)
    

Show the postISR:

# lsstCam : Running `stars_lsstCam_PS1_high_focal.inst` on hyak with `qckBkgndPert00.cmd` 


In [2]:
import run_ps1_functions as func
run=1
for instrument in  ['lsstCam']:
    for field in ['high']:
        for position in ['focal',]:
            for pert in ['00',]:
                cmd_file = f'qckBkgndPert{pert}.cmd'
                obshistid = func.calculate_obshistid(instrument,field,position,cmd_file,run)
                print(obshistid, ':', instrument, field, position, cmd_file, )

9010021 : lsstCam high focal qckBkgndPert00.cmd


copied manually:
    
    cp stars_lsstCam_PS1_high_extra_9010121.inst stars_lsstCam_PS1_high_focal_9010021.inst 
    
And doing

    nano stars_lsstCam_PS1_high_focal_9010021.inst
    
updated obshistid, updated opsim_rawseeing to 0.69 , removed move 10 ( to make it focal catalog)



Made a slurm submit script:  `runSlurm.sl` , containing: 
    
    
#!/bin/bash -l
#SBATCH --account astro
#SBATCH --partition astro
#SBATCH --nodes 11
#SBATCH --ntasks-per-node 20
#SBATCH -t 600:00:00
#SBATCH --job-name lsstHqf
#SBATCH --mem=246G  # 256GB is the smallest astro node , so if that's the only available select 246G here
##SBATCH --mem-per-cpu=8GB # Request 4GB RAM per core
# Specify the real memory required per node

##turn on e-mail notification
#SBATCH --mail-type=ALL
#SBATCH --mail-user=suberlak@uw.edu

## export all your environment variables to the batch job session
#SBATCH --export=all


## may not need that one, since everywhere I provide abspath ...
## specify working directory for the  job ...

#SBATCH --chdir=/gscratch/astro/suberlak/aos/
echo "starting at `date` on `hostname`"

pwd

python /gscratch/astro/suberlak/aos/AOS/run_ps1_phosim.py --instruments lsstCam --fields high --positions focal --root_dir /gscratch/astro/suberlak/aos/AOS/DM-28360 --cmd_file q
ckBkgndPert00_hyak.cmd --phosim_path /gscratch/astro/brycek/ts_repos/phosim_syseng4/phosim.py --phosim_p 220

echo "ended at `date` on `hostname`"



With https://wiki.cac.washington.edu/display/hyakusers/Maintenance+schedule I see that the  the nearest maintenance is  on Tuesday July 6. I find from here 

I check with this website https://howlongagogo.com/date/2021/july/6 that it's about 626 hours from now (Wed, Jun 9, right after the monthly maintenance...).


So I select `#SBATCH -t 600:00:00`


And with  

  
  sinfo -p astro 
  
I found there were 11 available nodes...



In [ ]:
And it is running ! 

    squeue -p astro 
    
shows the jobid `1970789`... 

# Simulate just the corner sensors with PS1 - submit to slurm or HTCondor ... 

stars_lsstCam_PS1_high_focal_9010021.instUse `stars_lsstCam_PS1_high_focal_9010021.inst`, so that the corner sensors are out-of-focus with proper distance ...

In [4]:
run=1
instrument='lsstCam'
field = 'high'
position='focal'
pert='00' # small pert 
#cmd_file = f'qckBkgndPert{pert}.cmd' # quick background 
cmd_file = f'noBkgndPert{pert}.cmd' # quick background 



obshistid = calculate_obshistid(instrument,field,position,cmd_file,run)
print(obshistid, ':', instrument, field, position, cmd_file, )

9010001 : lsstCam high focal noBkgndPert00.cmd


In [ ]:
Develop dry_run and specifying which sensors to simulate on  LSP:
    
    /home/scichris/notebooks/run_ps1_phosim.py
    
Add to github, push, change to

    cd /project/scichris/aos/AOS/
    git pull
    
    

submit on  lsst-devl03 , since slurm is all booked : 


    cd /home/scichris/notebooks/
    python run_ps1_phosim.py --positions focal --fields high --instruments lsstCam --phosim_p 24 --cmd_file qckBkgndPert00_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/ --phosim_s  R40_SW0 R40_SW1 --dry_run 



    python /project/scichris/aos/AOS/run_ps1_phosim.py --positions focal --fields high --instruments lsstCam --phosim_p 24 --cmd_file qckBkgndPert00_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/corner/ --dry_run 

First, change the SIM_VISTIME to 0.5 , so that I can see if the code performs well (and run with no background to get it faster too):
    
    cd /project/scichris/aos/AOS/DM-28360/
    nano stars_lsstCam_PS1_high_focal_9010021.inst # 0.5 sec, qck Bkgnd, Pert00 
    
    nano stars_lsstCam_PS1_high_focal_9010001.inst # 0.5 sec,  no Bkgnd, Pert00
    
    R00_SW0 R00_SW1 
    

In [17]:
detectorNames = ['R40_SW0', 'R40_SW1',
                         'R44_SW0', 'R44_SW1',
                         'R00_SW0', 'R00_SW1',
                         'R04_SW0', 'R04_SW1'
                    ]
detectors= ''
for detector in detectorNames:
    detectors += f" {detector}"

In [18]:
detectors

' R40_SW0 R40_SW1 R44_SW0 R44_SW1 R00_SW0 R00_SW1 R04_SW0 R04_SW1'

Run on lsst-devl03 in screen "closeLoop" , using 24 cores: 

     python /project/scichris/aos/AOS/run_ps1_phosim.py --positions focal --fields high --instruments lsstCam --phosim_p 24 --cmd_file noBkgndPert00_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/ --phosim_s  R40_SW0 R40_SW1 --dry_run 

Inspect the repackaged and post-ISR results : 

moved to `lsstCam_corner_short` 

## Ps1 high only corner: 

Running the full 15-sec, quick background, 4-corner sensors simulation, on lsst-devl03 screen "closeLoop" : 
    
    nano stars_lsstCam_PS1_high_focal_9010021.inst # 0.5 sec, qck Bkgnd, Pert00 
    
    --> change SIM_VISTIME to 15 sec 
    
    python /project/scichris/aos/AOS/run_ps1_phosim.py --positions focal --fields high --instruments lsstCam --phosim_p 24 --cmd_file qckBkgndPert00_NCSA.cmd --root_dir /project/scichris/aos/AOS/DM-28360/ --phosim_s  R40_SW0 R40_SW1 R44_SW0 R44_SW1 R00_SW0 R00_SW1 R04_SW0 R04_SW1  --dry_run
    
    
    phosim_repackager.py /project/scichris/aos/AOS/DM-28360/lsstCam/high/focal/9010021/raw     --out_dir /project/scichris/aos/AOS/DM-28360/lsstCam/high/focal/9010021/repackaged --inst lsst --focusz 0
    
    DONE ✅ 
    

## PS1 med only corner 

Running the full 15-sec, quick background, small pert (00),  on lsst-devl02 , screen pert  

     python /project/scichris/aos/AOS/run_ps1_phosim.py --positions focal --fields med --instruments lsstCam --phosim_p 24 --cmd_file qckBkgndPert00_NCSA.cmd  --root_dir /project/scichris/aos/AOS/DM-28360/ --phosim_s  R40_SW0 R40_SW1 R44_SW0 R44_SW1 R00_SW0 R00_SW1 R04_SW0 R04_SW1 --dry_run
    
    DONE ✅ 

## PS1 low only corner 

Running the full 15-sec, quick background, small pert (00),  on  lsst-devl03, screen  ps1 

source "/software/lsstsw/stack/loadLSST.bash"    # need that for python 
setup lsst_distrib 

source /project/scichris/aos/setup_aos_current.sh
setup -k -r /project/scichris/aos/phosim_utils/


     python /project/scichris/aos/AOS/run_ps1_phosim.py --positions focal --fields low --instruments lsstCam --phosim_p 24 --cmd_file qckBkgndPert00_NCSA.cmd  --root_dir /project/scichris/aos/AOS/DM-28360/ --phosim_s  R40_SW0 R40_SW1 R44_SW0 R44_SW1 R00_SW0 R00_SW1 R04_SW0 R04_SW1     ( --dry_run  to test if it's using the correct inst files ... ) 
    
    DONE ✅ 